infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/plugins/DefaultClusterManager.java

```diff
@@ -182,199 +204,339 @@ public TopologyDescriptor readConfig() {
-        ClusterDescriptor currentSource = currentConfig.getSourceClusters().values().iterator().next();
 
-        List<ClusterDescriptor> newSinkClusters = new ArrayList<>(currentConfig.getSinkClusters().values());
-        ClusterDescriptor newSink = new ClusterDescriptor(currentSource, ClusterRole.SINK);
-        newSinkClusters.add(newSink);
+        Map<String, ClusterDescriptor> newSinkClusters = new HashMap<>();
+        topologyConfig.getAllClustersInTopology().values().forEach(cluster -> {
+            if (DefaultClusterConfig.getSinkClusterIds().contains(cluster.getClusterId())) {
+                newSinkClusters.put(cluster.getClusterId(), cluster);
+            } else {
+                newSinkClusters.put(cluster.getClusterId(), cluster);
+            }
+        });
 
-        return new TopologyDescriptor(++configId, new ArrayList<>(), newSinkClusters, localEndpoint);
+        return new TopologyDescriptor(++configId, new HashMap<>(), new HashMap<>(), newSinkClusters,
+                new HashSet<>(), localNodeId);
     }
 
     /**
      * Create a new topology config, which marks all sink cluster as invalid on purpose.
      * LR should not replicate to these clusters.
      **/
     public TopologyDescriptor generateConfigWithInvalid() {
-        TopologyDescriptor currentConfig = new TopologyDescriptor(topologyConfig, localEndpoint);
-
-        List<ClusterDescriptor> newSourceClusters = new ArrayList<>(currentConfig.getSourceClusters().values());
-        List<ClusterDescriptor> newInvalidClusters = new ArrayList<>();
-        currentConfig.getSinkClusters().values().forEach(sinkCluster ->
-                newInvalidClusters.add(new ClusterDescriptor(sinkCluster, ClusterRole.INVALID)));
+        //Since all the sinks are marked as invalid, the remote source/sink/connectionEndPoints will all be empty as
+        //(i) if local cluster is source, there are no sinks and hence no connectionEndPoints
+        //(ii) if the local cluster is INVALID (previously sink), no LR against it.
+
+        Map<String, ClusterDescriptor> newInvalidClusters = new HashMap<>();
+        topologyConfig.getAllClustersInTopology().values().forEach(cluster -> {
+            if (DefaultClusterConfig.getSinkClusterIds().contains(cluster.getClusterId())) {
+                newInvalidClusters.put(cluster.getClusterId(), cluster);
+            } else {
+                newInvalidClusters.put(cluster.getClusterId(), cluster);
+            }
+        });
 
-        return new TopologyDescriptor(++configId, newSourceClusters, new ArrayList<>(), newInvalidClusters, localEndpoint);
+        return new TopologyDescriptor(++configId, new HashMap<>(), new HashMap<>(), newInvalidClusters,
+                new HashSet<>(), localNodeId);
     }
 
-    /**
-     * Bring topology config back to default valid config.
-     **/
     public TopologyDescriptor generateDefaultValidConfig() {
```

Comment:
can you rename this method to reflect its behavior better?  Something like generateSingleSourceSinkConfig()?

Reply:
I removed this and updateDefaultValidConfig() and instead made generateSingleSourceSinkTopolgy() public.